KMean Algorithm on iris data-set 

In [1]:
from sklearn import datasets
from sklearn.cluster import KMeans
from sklearn.utils import shuffle

import numpy as np
import pandas as pd

In [2]:
iris=datasets.load_iris()
X=iris.data
Y=iris.target
X,Y = shuffle(X,Y)

In [3]:
#Defining model 
model=KMeans(n_clusters=3,init='k-means++',max_iter=10,n_init=1,verbose=0,random_state=3425)

In [4]:
model.fit(X)
Y_Pred=model.labels_
Y_Pred

array([1, 2, 0, 0, 1, 2, 2, 0, 0, 1, 0, 0, 0, 0, 0, 2, 0, 0, 0, 2, 0, 1,
       1, 2, 0, 0, 0, 2, 2, 2, 1, 1, 0, 0, 0, 2, 2, 0, 1, 2, 2, 0, 2, 0,
       2, 0, 0, 0, 1, 2, 1, 1, 1, 1, 1, 1, 2, 1, 0, 1, 2, 1, 1, 0, 0, 0,
       1, 1, 0, 1, 0, 0, 1, 1, 2, 1, 2, 0, 1, 1, 0, 1, 0, 1, 2, 1, 0, 2,
       1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 2, 1, 2, 1, 1, 2, 2, 2,
       2, 0, 0, 1, 1, 2, 2, 1, 1, 0, 2, 0, 0, 1, 0, 2, 0, 0, 1, 2, 1, 1,
       1, 0, 1, 0, 2, 1, 0, 1, 2, 0, 2, 2, 2, 0, 0, 0, 2, 0], dtype=int32)

In [5]:
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(Y,Y_Pred)
print(cm)
from sklearn.metrics import accuracy_score
print(accuracy_score(Y,Y_Pred))

[[ 0 50  0]
 [47  0  3]
 [14  0 36]]
0.24


EM Algorithm

In [6]:
iris = datasets.load_iris()
X = pd.DataFrame(iris.data)
Y = pd.DataFrame(iris.target)

In [7]:
from sklearn.mixture import GaussianMixture
model2=GaussianMixture(n_components=3,random_state=3425)
model2.fit(X)

GaussianMixture(n_components=3, random_state=3425)

In [8]:
#Predicting classes for our data
uu= model2.predict(X)
#Accuracy of EM Model
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(Y,uu)
print(cm)
from sklearn.metrics import accuracy_score
print(accuracy_score(Y,uu))

[[ 0  0 50]
 [45  5  0]
 [ 0 50  0]]
0.03333333333333333


Accuracy we get from KMean algorithm is 0.24 which is less than that of  accuracy we get from EM algorithm that is 0.33 on same data-set (iris data-set)